File: LeetCode/Top K Frequent Elements/Top K Frequent Elements.cpp
// https://leetcode.com/problems/top-k-frequent-elements/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        ios::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        if(nums.size() == 0 or nums.size() == 1){
            return nums;
        }

        unordered_map<int, int> map1; // Numero: conteo de veces que ha apareceido
        unordered_map<int, vector<int>> map2; // Numero de veces que se repite : numeros correspondientes
        vector<int> result;

        for(int i = 0; i < nums.size(); i++){
            int n, c;
            n = nums[i];
            map1[n]++;
            c = map1[n];
            map2[c].push_back(n);
        }
        
        for(int i = nums.size() - 1; i > 0; i--){
            vector<int> v = map2[i];
            for(auto x: v){
                if(find(result.begin(), result.end(), x) == result.end()){ // Verifica si x esta en result
                    result.push_back(x);
                }
                if(result.size() == k){
                    break;
                }
            }
            if(result.size() == k){
                break;
            }
        }

        return result;
    }
};

int main(){
    Solution Solucion;

    vector<int> nums = {3,2,3,1,2,4,5,5,6,7,7,8,2,3,1,1,1,10,11,5,6,2,4,7,8,5,6}; 
    int k = 3;

    vector<int> result = Solucion.topKFrequent(nums, k);
    for(auto x: result){
        cout << x << " ";
    }
}

File: LeetCode/Top K Frequent Elements/Top_K_Frequent_Elements.py
# https://leetcode.com/problems/top-k-frequent-elements/
class Solution:
    def topKFrequent(self, nums, k):
        if len(nums) == k:
            return nums

        freq_nums = dict() # key: number, value: frequency

        for num in nums:
            if num not in freq_nums:
                freq_nums[num] = 1
            else:
                freq_nums[num] = freq_nums[num] + 1
        
        frequences = list(set(freq_nums.values()))
        frequences.sort(reverse=True)
        
        freq = dict() # key: frequency, value: list of numbers
        for key, value in freq_nums.items():
            if value in freq:
                freq[value].append(key)
            else:
                freq[value] = [key]

        resullt = []
        for f in frequences:
            if len(resullt) < k:
                resullt.extend(freq[f])
            else:
                break
        return resullt[:k]
    
Solution = Solution()

nums = [3,2,3,1,2,4,5,5,6,7,7,8,2,3,1,1,1,10,11,5,6,2,4,7,8,5,6]
k = 10
print(Solution.topKFrequent(nums, k))